### ISSUE TITLE
Darc incorrectly constructing download url for internal manifest symbols package

### ISSUE THREAD

**Issue by mmitche** [dotnet/arcade#7729]
- [ ] This issue is blocking 
- [ ] This issue is causing unreasonable pain

https://dev.azure.com/dnceng/internal/_build/results?buildId=1282580&view=logs&j=40e1e683-ea82-5be9-b25e-899b65e1c0cb&t=94a692ec-20f2-55dd-1884-cf0cad1763a4&l=7141

This is an internal build, so the symbol package was uploaded to a nuget feed. The format of the package name appears to be confusing darc's gather-drop, because it's incorrectly constructing the package download URL:

It attempts `https://pkgs.dev.azure.com/dnceng/internal/_apis/packaging/feeds/dotnet6-internal-symbols/nuget/packages/Microsoft.NET.Workload.Mono.ToolChain.Manifest-6.0.100.Msi.x64.6.0.0-preview.7.21377.20/versions/6.0.100/content`. This appears to mean that it thinks the version number is `6.0.100`, rather than `6.0.0-preview.7.21377.20`

My hunch is that the - in the middle of the package name is confusing things. It's unclear whether the error is on the arcade-services side or the version identification side. BAR appears to understand the version number correctly, according to the darc get-asset output, but it's unclear why it would choose 6.0.100 later.

```
PS C:\r\arcade> darc get-asset --name "assets/symbols/Microsoft.NET.Workload.Mono.ToolChain.Manifest-6.0.100.6.0.0-preview.7.21377.20.symbols.nupkg"
Looking up assets with name 'assets/symbols/Microsoft.NET.Workload.Mono.ToolChain.Manifest-6.0.100.6.0.0-preview.7.21377.20.symbols.nupkg' in the last 30 days
assets/symbols/Microsoft.NET.Workload.Mono.ToolChain.Manifest-6.0.100.6.0.0-preview.7.21377.20.symbols.nupkg @ 6.0.0-preview.7.21377.20
Repository:    https://dev.azure.com/dnceng/internal/_git/dotnet-runtime
Branch:        refs/heads/internal/release/6.0-preview7
Commit:        01019e12c1292fc51496c2c3b52fb5141d0f0602
Build Number:  20210727.20
Date Produced: 7/27/2021 9:03 PM
Build Link:    https://dev.azure.com/dnceng/internal/_build/results?buildId=1262709
BAR Build Id:  99704
Released:      False
Channels:
- .NET 6 Preview 7 Internal
Locations:
- https://dev.azure.com/dnceng/internal/_apis/build/builds/1262709/artifacts (Container)
- https://pkgs.dev.azure.com/dnceng/internal/_packaging/dotnet6-internal-symbols/nuget/v3/index.json (Container)
```

**Comment by MattGal** [comment]
[Triage] :  Seems like a need-to-fix issue and no obvious epic for it, @ilyas1974 FR?

**Comment by ilyas1974** [comment]
I think we can take a "first pass" at this in FR.

**Comment by mmitche** [comment]
> I think we can take a "first pass" at this in FR.

Yeah I think this will turn out to be a simple fix in gather-drop's code.

**Comment by adiaaida** [comment]
I can take this issue, as I've got context on gather-drop.

**Comment by adiaaida** [comment]
Matt, in your sample, where it has the version correctly, I think you looked up the wrong package. The one with the malformed version is `Microsoft.NET.Workload.Mono.ToolChain.Manifest-6.0.100.Msi.x64.6.0.0-preview.7.21377.20` and you looked up `Microsoft.NET.Workload.Mono.ToolChain.Manifest-6.0.100.6.0.0-preview.7.21377.20` (note the missing .Msi). When I looked up the Msi package in BAR, I found this:

<img width="1260" alt="Screen Shot 2021-08-18 at 4 39 38 PM" src="https://user-images.githubusercontent.com/9666644/129985665-77606990-a904-4010-9445-8d5e35ce6a7c.png">

It is definitely in BAR with the 6.0.100 version, suggesting that gather-drop is doing the right thing, but elsewhere where we publish (?) the file to BAR, it doesn't get the version correctly.

**Comment by adiaaida** [comment]
Does anyone know where this package comes from?

**Comment by MattGal** [comment]
> Does anyone know where this package comes from?

https://github.com/dotnet/runtime/blob/main/src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/Microsoft.NET.Workload.Mono.Toolchain.Manifest.pkgproj I think

**Comment by adiaaida** [comment]
Thanks Matt. I'm search for the azdo build to see if I can find anything obvious there.

**Comment by adiaaida** [comment]
Yeah, the problem is definitely somewhere in our algorithm for determining versions https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs#L82

**Comment by adiaaida** [comment]
Specifically

```
// - There have been at least 3 numbers found and the current segment is not a number or not preceded by '-'. In this case, we can call this the major minor
//   patch number and no longer need to continue searching
```

That's what's happening in this case: we are getting 6.0.100, the next delimiter is a . and the next segment is Msi, so we stop looking for the version. I'm not 100% sure what the fix should be here: either we ask the mono folks to remove the SdkBandVersion from the name of their packages (that's where the 6.0.100 is coming from -- https://github.com/dotnet/runtime/blob/main/src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/Microsoft.NET.Workload.Mono.Toolchain.Manifest.pkgproj#L11), or we do some work to fix the algorithm for finding the version. I'm not sure which is best, but certainly easiest would be to have them remove it from their package name (assuming they don't have a good reason for SdkBandAndVersion in the name of the package).

**Comment by adiaaida** [comment]
Ok, I think I have a change that might fix the version parser. I will put up a PR and we can discuss.

**Comment by mmitche** [comment]
The sdk band version has to be in the package name, since there will eventually be multiple of them. I think we either need to change the version identifier (with a new test of course, since this stuff can be fragile) or change the way the package name is constructed so that it still contains 6.0.100, but perhaps with different delimiters. That may require reaction on the VS or installer side.

/cc @joeloff

**Comment by adiaaida** [comment]
PR: https://github.com/dotnet/arcade/pull/7769

**Comment by adiaaida** [comment]
For the "change the delimiter" idea, the algorithm treats all of our accepted delimiters (`-`, `.`, `_`) the same, so changing the delimiter doesn't fix the algorithm's problem.

**Comment by mmitche** [comment]
This appears to be fixed in the arcade tests, but somehow hasn't shown up in the runtime builds yet. We did flow the fix into arcade-services (I think that's required so that the push to BAR task picks it up). But perhaps a new push to bar task needs to make it back into arcade? 

https://dev.azure.com/dnceng/internal/_build/results?view=results&buildId=1377024

The binlog for the Publish To Build Asset Registry job is probably a good place to look first, checking the version of the loaded assembly that contains the version identification task.

**Comment by MattGal** [comment]
Put back to the backlog and unassigned so we can get a fresh investigation on this.  With luck it's just a need for dependency updates to flow.

**Comment by epananth** [comment]
I checked the MicrosoftDotNetMaestroTasksVersion in arcade and its currently using 1.1.0-beta.21378.2, which was updated on August 3rd. Michelle's changes went in to arcade on August 19th. Maestro task version has no been updated since August 3rd. The latest version of Maestro task is -> 1.1.0-beta.21479.2, https://dev.azure.com/dnceng/internal/_build/results?buildId=1393447&view=logs&j=3f8da9d0-12ef-55d8-ee3f-b7a2a2f0694c&t=ab89c890-dd91-5106-cf69-373a0066ef78&l=93

My plan is to update this version and run a test against Runtime.

**Comment by epananth** [comment]
The version update has been propagated to main. Closing this one
